smoke_tests/tools/package/base_builder.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2014-2020 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import absolute_import
+
+if False:
+    from typing import Optional
+
+from abc import ABCMeta, abstractmethod
+
+import six
+
+from smoke_tests.tools.utils import (
+    create_temp_dir_with_constant_name,
+    copy_agent_source as copy_source,
+)
+from smoke_tests.tools.compat import Path
+
+
+@six.add_metaclass(ABCMeta)
+class AgentImageBuilder(object):
+    """
+     Abstraction to build docker images.
+    """
+
+    IMAGE_TAG = None  # type: six.text_type
+
+    # add agent source code to the build context of the image
+    COPY_AGENT_SOURCE = False  # type: bool
+
+    def __init__(self):
+        self._docker = None  # type: Optional
+
+    @property
+    def _docker_client(self):
+        if self._docker is None:
+            import docker
+
+            self._docker = docker.from_env()
+
+        return self._docker
+
+    @property
+    def image_tag(self):  # type: () -> six.text_type
+        return type(self).IMAGE_TAG
+
+    @property
+    def _copy_agent_source(self):  # type: () -> bool
+        return type(self).COPY_AGENT_SOURCE
+
+    @classmethod
+    @abstractmethod
+    def get_dockerfile_content(cls):  # type: () -> six.text_type
+        """
+        Get the content of the Dockerfile.
+        """
+        pass
+
+    def build(self):
+        """
+        Build docker image.
+        """
+        build_context_path = create_temp_dir_with_constant_name(".scalyr_agent_testing")
+
+        dockerfile_path = build_context_path / "Dockerfile"
```

Comment:
Should use os.path.join

Reply:
I thought this operator does the same thing https://docs.python.org/3/library/pathlib.html#operators

Reply:
Ah, sorry, I didn't know you are working with PathLib objects, I assume you were working with plain strings :)